File: problems/building_h2o/solution.py
import threading
class H2O(object):
    def __init__(self):
        self.lock = threading.Lock()
        self.hnum = 0
        self.onum = 0
        self.hrelease = None
        self.orelease = None

    def hydrogen(self, releaseHydrogen):
        self.hrelease = releaseHydrogen
        self.output(1, 0)

    def oxygen(self, releaseOxygen):
        self.orelease = releaseOxygen
        self.output(0, 1)

    def output(self, h, o):
        with self.lock:
            self.hnum += h
            self.onum += o
            if self.hnum >= 2 and self.onum >= 1:
                self.hrelease()
                self.hrelease()
                self.orelease()
                self.hnum -= 2
                self.onum -= 1

File: problems/building_h2o/solution.cpp
using namespace std;
class H2O {
    condition_variable cv;
    mutex mu;
    int h, o;
public:
    H2O() {
        h = o = 0;
    }

    void hydrogen(function<void()> releaseHydrogen) {
        unique_lock<mutex> lg(mu);
        cv.wait(lg, [this]{return this->h != 2;});
        h ++;
        releaseHydrogen();
        form();
        cv.notify_all();
    }
    
    void form() {
        if (o > 0 and h >= 2 * o) {
            o --;
            h -= 2;
        }
    }

    void oxygen(function<void()> releaseOxygen) {
        unique_lock<mutex> lg(mu);
        cv.wait(lg, [this]{return this->h >= (this->o + 1) * 2;});
        o ++;
        releaseOxygen();
        form();
        cv.notify_all();
    }
};